### PR TITLE
dump debug functions for debugging and Lua development

### PR DESCRIPTION
**Pull request by Znote**
Adding my favourite helper dump function to a new data/lib/debugging folder and loading it. 

I find these functions very useful while debugging Lua code scripts, especially if something goes wrong and I need a closer look at the current state of Lua tables in-game. I find myself manually including them on a daily basis when coding Lua for various reasons. 

They are especially handy if the table is dynamic and continuously populated and manipulated live. Instead of reading potentially many lines of code and guessing how the table is manipulated, rows inserted and deleted etc, I can just dump it to console whenever I want and look at the information. 

Creating this pull request in case someone else might find them handy, having it added also alleviate me grabbing them from gist and pasting them everywhere. x) 

dump(input): 
- Convert input data to human readable string

pdump(input) 
- execute dump(input) and print it to console. (print the dump)

tdump(title, input): 
- Beautify the output further by adding newlines before and after dump, as well as adding a title to the console to represent this data. (print the dump better using a title)

The function names are short and obscure on purpose, once you get used to them, it is nice to be able to quickly and effortlessly use them. 

Demo code:
```lua
local talkaction = TalkAction("!hello")

function talkaction.onSay(player)
   print(player:getName() .. ' executed talkaction: !hello')
    -- SAMPLE START

   local spell = Spell("Strong Haste")
   local info = {
      ["name"] = player:getName(),
      ["level"] = player:getLevel(),
      ["position"] = player:getPosition(),
      ["town_name"] = player:getTown():getName(),
      ["town_id"] = player:getTown():getId(),
      ["players"] = Game.getPlayers(),
      ["spells"] = {
         [spell:name()] = {
            ["words"] = spell:words(),
            ["vocations"] = spell:vocation()
         }
      }
   }

   -- print the dump results with a beautified title 
   tdump('Player Information', info)

   -- Just print the dump quickly without any mess
   pdump(info)

   -- Just grab the dump string for further manipulation.
   local dump_str = dump(info) 
   print(dump_str)

   -- SAMPLE END
   return true
end
```

Looks like this in console:
```lua
>> Loading map
> Map size: 1000x1000.
> Map loading time: 0.87 seconds.
> Loaded house items in: 0.001 s
>> Initializing gamestate
>> Loaded all modules, server starting up...
>> Forgotten Server Online!

Luxitur has logged in.
Luxitur has logged out.
Luxitur has logged in.
Luxitur executed talkaction: !hello

==========================
=== Player Information ===
==========================
{
    ["level"] = 90,
    ["spells"] = {
        ["Strong Haste"] = {
            ["vocations"] = {
                ["2"] = "Druid",
                ["4"] = "Elder Druid",
                ["1"] = "Sorcerer",
                ["3"] = "Master Sorcerer"
            },
            ["words"] = "utani gran hur"
        }
    },
    ["town_name"] = "Rhyves",
    ["position"] = {
        ["y"] = 363,
        ["x"] = 109,
        ["z"] = 6,
        ["stackpos"] = 0
    },
    ["town_id"] = 2,
    ["name"] = "Luxitur",
    ["players"] = {
        [1] = userdata: 0x7fd28388
    }
}
==========================

Luxitur has logged out.
Luxitur has logged in.
Luxitur has logged out.
Luxitur has logged in.
```

Notice that the dump output (with exception of userdata / function references) can be copy-pasted directly back into Lua environment (Lua table definition compliant) in a variable to if you need to simulate the state and test things out. 

If the dump is small enough for a player message, you could use `dump(input)` to return the string without printing it to a console, and print it out to the player instead. (etc the player who triggers a talkaction). Alternatively, use `pdump(input)` and do both.

### CODE REVIEW

File: data/lib/debugging/dump.lua [hunk 1]
```diff
@@ -0,0 +1,51 @@
+-- recursive dump function
+function dumpLevel(input, level)
+   local indent = ''
+   for i = 1, level do
+      indent = indent .. '    '
```
Comment (WibbenZ) [inline data/lib/debugging/dump.lua]: Isn't \t better here to print out a tab?

File: data/lib/debugging/dump.lua [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+   end
+
+   if type(input) == 'table' then
+      local str = '{ \n'
+      local lines = {}
+      for k,v in pairs(input) do
+         if type(k) ~= 'number' then k = '"'..k..'"' end
```
Comment (WibbenZ) [inline data/lib/debugging/dump.lua]: Looks better if you add a space between k,v aswell as indent the if statments.

File: data/lib/debugging/dump.lua [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+         if type(v) == 'string' then v = '"'..v..'"' end
```
Comment (WibbenZ) [inline data/lib/debugging/dump.lua]: Spaces before and after v

File: data/lib/debugging/dump.lua [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+         table.insert(lines, indent .. '    ['..k..'] = ' .. dumpLevel(v, level+1))
```
Comment (WibbenZ) [inline data/lib/debugging/dump.lua]: Spaces before and after +

File: data/lib/debugging/dump.lua [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	header_str = header_str .. '=== ' .. title .. ' ===\n'
+	header_str = header_str .. '====' .. title_fill .. '====\n'
+
+	local dump_str = dump(input)
+	local footer_str = '\n====' .. title_fill .. '====\n'
+
+	print(header_str .. dump_str .. footer_str)
+
+	return dump_str
+end
```
Comment (WibbenZ) [inline data/lib/debugging/dump.lua]: Missing EOF new line

File: data/lib/lib.lua [hunk 6]
```diff
@@ -3,3 +3,6 @@ dofile('data/lib/core/core.lua')
 
 -- Compatibility library for our old Lua API
 dofile('data/lib/compat/compat.lua')
+
+-- Debugging helper function for Lua developers
+dofile('data/lib/debugging/dump.lua')
```
Comment (WibbenZ) [inline data/lib/lib.lua]: Missing EOF new line

File: data/lib/debugging/dump.lua [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      end
+      return str .. table.concat(lines, ',\n') .. '\n' .. indent .. '}'
+   end
+
+   return tostring(input)
+end
+
+-- Return a string representation of input for debugging purposes
+function dump(input)
+   return dumpLevel(input,0)
```
Comment (WibbenZ) [inline data/lib/debugging/dump.lua]: Missing a space after input,

File: data/lib/debugging/dump.lua (hunk 1 above)
Comment (Znote) [reply data/lib/debugging/dump.lua]: `\t` often defaults to 8 spaces which I dislike. It also looks aweful in-game when sending as a text message. 

First pdump is `\t`, second is just 4 blank spaces:
image

Comment (WibbenZ) [reply data/lib/debugging/dump.lua]: True, looks better :)

**Comment by Znote** [comment]
> A nice thing to use when debugging, but do we need it in the repo?
> At the same time who would check OTLand for debugging code / scripts, question is how people will find the code, mixed in a lib file or on OTLand.

These functions are not essential, so they don't have to find them. :P But I imagine helping people out in support board, it is easier to just tell them to "add `tdump("WibbenZ Table of curiosity", achievements)`" in line 34 and tell us the results, instead of asking them to implement a debugging function, then use it properly, then tell us the results. 

So I imagine this is handy not only during development, but during support cases on otland. Either way, for those of us who knows about it, it saves time and is a convenient thing. In my opinion, we can handle a few tiny functions as bloat for the convenience of developers. This isn't something that slows the server down for those ignorant of the feature.

**Comment by Znote** [comment]
I havent done it yet, but I can also imagine that these functions are very useful for OT specific Lua tutorials, as it gives the instructor, and the student a clear method of presenting table data.

**Pull request merged**